### PR TITLE
Add RSA key validation to default provider

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
EVP_PKEY_CTX_new_from_pkey() is added as part of another PR - it is only added here for testing.

**Comment by slontis** [comment]
Rebased..

File: crypto/rsa/rsa_chk.c [hunk 1]
```diff
@@ -236,5 +221,40 @@ int RSA_check_key_ex(const RSA *key, BN_GENCB *cb)
...
+{
+    return RSA_check_key_ex(key, NULL);
+}
+
+int RSA_check_key_ex(const RSA *key, BN_GENCB *cb)
+{
+#ifdef FIPS_MODE
+    return rsa_validate_public(key)
+           && rsa_validate_private(key)
+          && rsa_validate_pairwise(key);
```
Comment (levitte) [inline crypto/rsa/rsa_chk.c]: Zigzag indent? :wink:

File: doc/man7/provider-keymgmt.pod [hunk 2]
```diff
@@ -43,6 +43,13 @@ provider-keymgmt - The KEYMGMT library E<lt>-E<gt> provider functions
  /* Discovery of supported operations */
  const char *OP_keymgmt_query_operation_name(int operation_id);
 
+/* Key validation */
+
+int OP_keymgmt_validate_domparams(void *key);
```
Comment (levitte) [inline doc/man7/provider-keymgmt.pod]: Nit: "key" should be "domparams", at least if you want to be consistent with the rest

File: providers/implementations/keymgmt/rsa_kmgmt.c [hunk 3]
```diff
@@ -244,11 +244,35 @@ static const OSSL_PARAM *rsa_importkey_types(void)
     return rsa_key_types;
 }
 
+static int rsa_validatekey_public(void *key)
```
Comment (levitte) [inline providers/implementations/keymgmt/rsa_kmgmt.c]: Done forget to add forward declarations at the top of this file, so the compiler can help out confirming that you got the signature right...

**Comment by levitte** [comment]
I like this :-)

Need to make second read, though

File: doc/man7/provider-keymgmt.pod (hunk 2 above)
Comment (slontis) [reply doc/man7/provider-keymgmt.pod]: shifted into its own section

**Comment by slontis** [comment]
rebased

**Comment by slontis** [comment]
rebased that pesky libcrypto.num again.

**Comment by slontis** [comment]
ping

**Comment by mattcaswell** [comment]
Although looks like a rebase is in order.

**Comment by slontis** [comment]
rebased - probably needs a reapproval @mattcaswell

**Comment by slontis** [comment]
rebased again..

**Comment by slontis** [comment]
I would really like to know why build #4 is timing out quite often AFTER the tests have completed

**Comment by slontis** [comment]
trivial rebase of libcrypto.num

**Comment by slontis** [comment]
Yet another trivial rebase.

**Comment by slontis** [comment]
@mattcaswell can you change the approval label if you are happy with the rebase..
This is a good example of why waiting 24 hours after a rebase is just not practical.

**Comment by kroeckx** [comment]
I suggest that if it's a trivial change, because for instance the *.num
files or CHANGES conflicted, that you don't need to ask to reconfirm it.

**Comment by mattcaswell** [comment]
I consider this "ready to merge" so I have updated the label accordingly.

**Comment by slontis** [comment]
Thanks. Merged to master.

**Pull request closed without merging**